BUGFIX: Fix like queries using paths in NodeDataRepository

kdambekalns opened the pull request:
When querying for nodes based on (parent) paths, a LIKE query was used in
certain cases. This query would include unrelated nodes, if the paths involved
would share a common prefix:

    /some/node-12
    /some/node-123

Nodes below both paths would be included because of:

    LIKE "/some/nodes-12%"

Now those queries append a slash, to read:

    LIKE "/some/nodes-12/%"

NEOS-1849 #close

kdambekalns commented:
Maybe @inchie can check this…

Inchie commented:
I am afraid. But I have found one conflict. 
` $context = $this->contextFactory->create(['workspaceName' => 'live']);
 $context->getRootNode()
`
The root node has the parentPath of '/'. So you have on Line 1490 `->setParameter('parentPath', $parentPath . '/%');` two slashes in the query.

kdambekalns commented:
Right! In had added `rtrim(..., '/')` in a few places, but that broke `/` paths. So that indeed needs special handling. Thanks for testing!

> Am 17.06.2016 um 21:30 schrieb Timo Nußbaum <notifications@github.com>:
> 
> I am afraid. But I have found one conflict. 
> $context = $this->contextFactory->create(['workspaceName' => 'live']);
> $context->getRootNode()
> 
> The root node has the parentPath of '/'. So you have on Line 1490 ->setParameter('parentPath', $parentPath . '/%'); two slashes in the query.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Inchie commented:
You are welcome.

aertmann commented:
looks good, but should add test you be added

kitsunet commented:
Use `\TYPO3\TYPO3CR\Domain\Utility\NodePaths::getParentPath` pls :)

kitsunet commented:
Ah, but that was not yet in 2.0 :(

kdambekalns commented:
Fixed the `/` issue, added tests (found another issue through that, thanks @aertmann for insisting).

@kitsunet where should `getParentPath` be used (when doing the upmerge)?

kitsunet commented:
@kdambekalns nvm, I misunderstood the ``rtrim(..., '/')`` stuff to be a remake of getting the parent path.

Inchie commented:
Testing again. For me it works!
